src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/StreamHelper.cs

```diff
@@ -383,9 +383,24 @@ private static string StreamToString(Stream stream, Encoding encoding)
 
         internal static string DecodeStream(Stream stream, string characterSet, out Encoding encoding)
         {
-            bool isDefaultEncoding = !TryGetEncoding(characterSet, out encoding);
+            bool isDefaultEncoding = false;
+            byte[] buffer = new byte[4];
 
+            if (stream.Length >= 4)
+            {
+                stream.ReadExactly(buffer, 0, 4);
+            }
+
+            EncodingHelper.TryDetectEncodingFromBom(buffer, out encoding, out int preambleLength);
```

Comment:
I would be surprised if this is not already done in .Net. It might be worth looking at HttpClient and other related code.

Reply:
I read the code in .Net and it doesn't match perfectly with our needs (it doesn't consider UTF32-BE, it checks for the BOM after checking the charset)

Reply:
> it checks for the BOM after checking the charset

Maybe it is more right behavior? Could you please point the code?

Reply:
Gets encoding from charset:
https://github.com/dotnet/runtime/blob/bd6cbe3642f51d70839912a6a666e5de747ad581/src/libraries/System.Net.Http/src/System/Net/Http/HttpContent.cs#L194-L219

Gets encoding from bom:
https://github.com/dotnet/runtime/blob/bd6cbe3642f51d70839912a6a666e5de747ad581/src/libraries/System.Net.Http/src/System/Net/Http/HttpContent.cs#L221-L234

Reply:
The code is 8 years old. So we should follow the logic too.

Reply:
I followed what was discussed here: https://github.com/PowerShell/PowerShell/issues/11547#issuecomment-636522169

Reply:
Thanks for pointing the history. It is correct.

Can we use StreamReader with auto encoding detection?
https://source.dot.net/#System.Private.CoreLib/src/libraries/System.Private.CoreLib/src/System/IO/StreamReader.cs,109

Reply:
@iSazonov I tried using StreamReader with auto encoding detection, what do you think?

```cs
        internal static string DecodeStream(Stream stream, string characterSet, out Encoding encoding)
        {
            bool isDefaultEncoding = false;

            StreamReader reader = new(stream, detectEncodingFromByteOrderMarks: true);

            // reader.CurrentEncoding defaults to UTF8
            encoding = reader.CurrentEncoding;

            if (encoding == Encoding.UTF8)
            {
                isDefaultEncoding = !TryGetEncodingFromCharset(characterSet, out encoding);
            }

            if (isDefaultEncoding)
            {
                reader = new(stream, encoding);

                // We only look within the first 1k characters as the meta element and
                // the xml declaration are at the start of the document
                int bufferLength = (int)Math.Min(reader.BaseStream.Length, 1024);

                char[] buffer = new char[bufferLength];
                reader.ReadBlock(buffer, 0, bufferLength);

                string substring = new(buffer);

                // Check for a charset attribute on the meta element to override the default
                Match match = s_metaRegex.Match(substring);
                
                // Check for a encoding attribute on the xml declaration to override the default
                if (!match.Success)
                {
                    match = s_xmlRegex.Match(substring);
                }
                
                if (match.Success)
                {
                    characterSet = match.Groups["charset"].Value;

                    if (TryGetEncodingFromCharset(characterSet, out Encoding localEncoding))
                    {
                        encoding = localEncoding;
                    }
                }
            }

            return new StreamReader(stream, encoding).ReadToEnd();
        }
```

Reply:
I guess we can use https://source.dot.net/#System.Private.CoreLib/src/libraries/System.Private.CoreLib/src/System/IO/StreamReader.cs,119
and pass encoding based on characterSet. This will aromatically use BOM if present.